lib/grandpa/neighbor_tracker.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+// https://github.com/paritytech/polkadot-sdk/blob/08498f5473351c3d2f8eacbe1bfd7bc6d3a2ef8d/substrate/client/consensus/grandpa/src/communication/mod.rs#L73 //nolint
+const neighbourBroadcastPeriod = time.Minute * 2
+
+type neighborData struct {
+	peer        peer.ID
+	neighborMsg *NeighbourPacketV1
+}
+
+type neighborState struct {
+	setID            uint64
+	round            uint64
+	highestFinalized uint32
+}
+
+type neighborTracker struct {
+	sync.Mutex
+	grandpa *Service
+
+	peerview         map[peer.ID]neighborState
+	currentSetID     uint64
+	currentRound     uint64
+	highestFinalized uint32
+
+	finalizationCha chan *types.FinalisationInfo
+	neighborMsgChan chan neighborData
+	stoppedNeighbor chan struct{}
+}
+
+func newNeighborTracker(grandpa *Service, neighborChan chan neighborData) *neighborTracker {
+	return &neighborTracker{
+		grandpa:         grandpa,
+		peerview:        make(map[peer.ID]neighborState),
+		finalizationCha: grandpa.blockState.GetFinalisedNotifierChannel(),
+		neighborMsgChan: neighborChan,
+		stoppedNeighbor: make(chan struct{}),
+	}
+}
+
+func (nt *neighborTracker) Start() {
+	go nt.run()
+}
+
+func (nt *neighborTracker) Stop() {
+	nt.grandpa.blockState.FreeFinalisedNotifierChannel(nt.finalizationCha)
+	close(nt.stoppedNeighbor)
```

Comment:
You should wait on `nt.run()` to finish in this `Stop` method.

Reply:
Nice call, will add